### PR TITLE
[CARBONDATA-4210] Handle 3.1 parsing failures related to alter complex types

### PR DESCRIPTION
**Pull request by akkio-97**
### Why is this PR needed?
For 2.3 and 2.4 parsing of **alter** commands are done by spark. Which is not in the case of 3.1.

 ### What changes were proposed in this PR?
- So carbon is responsible for the parsing here. 
- Previously ignored test cases due to this issue are now enabled.

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5618/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3873/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/20/

**Comment by akkio-97** [comment]
retest this please

File: integration/spark/pom.xml [hunk 1]
```diff
@@ -646,6 +646,9 @@
     </profile>
     <profile>
       <id>spark-3.1</id>
+      <activation>
```
Comment (vikramahuja1001) [inline integration/spark/pom.xml:649]: revert these changes

File: pom.xml [hunk 2]
```diff
@@ -126,11 +126,15 @@
     <hadoop.version>2.7.2</hadoop.version>
```
Comment (vikramahuja1001) [inline pom.xml:126]: revert all changes in this class

File: integration/spark/pom.xml (hunk 1 above)
Comment (akkio-97) [reply integration/spark/pom.xml:649]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/parser/CarbonSparkSqlParserUtil.scala [hunk 3]
```diff
@@ -447,21 +447,31 @@ object CarbonSparkSqlParserUtil {
...
-      dataType: String,
+      dataType: Option[String],
       values: Option[List[(Int, Int)]],
-      comment: Option[String]
+      comment: Option[String],
+      complexChild: Option[Field]
   ): CarbonAlterTableColRenameDataTypeChangeCommand = {
     val isColumnRename = !columnName.equalsIgnoreCase(columnNameCopy)
-    val alterTableColRenameAndDataTypeChangeModel =
+    val alterTableColRenameAndDataTypeChangeModel = if (!dataType.equals(None)) {
```
Comment (vikramahuja1001) [inline integration/spark/src/main/scala/org/apache/spark/sql/parser/CarbonSparkSqlParserUtil.scala:456]: instead of duplicating the whole thing, just put the columnName/columnNameCopy and datatype.get.tolowerCase/complexChild to 2 variables depending on the conditions and use them, because most of the variables are same.

File: pom.xml (hunk 2 above)
Comment (akkio-97) [reply pom.xml]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/parser/CarbonSparkSqlParserUtil.scala (hunk 3 above)
Comment (akkio-97) [reply integration/spark/src/main/scala/org/apache/spark/sql/parser/CarbonSparkSqlParserUtil.scala]: done

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5619/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3874/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/21/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5625/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3882/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/29/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5628/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3885/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/32/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5632/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3889/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/36/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3898/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5641/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/45/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3909/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/56/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5652/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3913/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5657/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/60/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/69/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5665/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3921/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/100/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5696/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3953/

**Comment by kunal642** [comment]
LGTM
